lib/pause_2017/PAUSE/Web/Plugin/WithMFAProtection.pm

```diff
@@ -0,0 +1,86 @@
+package PAUSE::Web::Plugin::WithMFAProtection;
+
+use Mojo::Base 'Mojolicious::Plugin';
+
+our $VERSION = '1.00';
+
+sub register {
+    my ( $self, $app ) = @_;
+
+    my $routes = $app->routes;
+
+    $routes->add_condition(
+        with_mfa_protection => sub {
+            my ( $route, $c ) = @_;
+
+            my $u = $c->active_user_record;
+
+            # XXX: The active user record does not have mfa when an admin user is pretending someone else.
+            return 1 unless $u->{mfa};
+
+            my $otp = $c->req->body_params->param('otp');
+            if (defined $otp and $otp ne '') {
+                if ($otp =~ /\A[0-9]{6}\z/) {
+                    return 1 if $c->app->pause->authenticator_for($u)->verify($otp);
+                } elsif ($otp =~ /\A[a-z0-9]{5}\-[a-z0-9]{5}\z/) { # maybe one of the recovery codes?
+                    require PAUSE::Crypt;
+                    my $pause = $c->stash(".pause");
+                    my @recovery_codes = split / /, $u->{mfa_recovery_codes} // '';
+                    for my $code (@recovery_codes) {
+                        if (PAUSE::Crypt::password_verify($otp, $code)) {
+                            my $new_codes = join ' ', grep { $_ ne $code } @recovery_codes;
+                            my $dbh = $c->app->pause->authen_connect;
+                            my $tbl = $PAUSE::Config->{AUTHEN_USER_TABLE};
+                            my $sql = "UPDATE $tbl SET mfa_recovery_codes = ?, changed = ?, changedby = ? WHERE user = ?";
+                            $dbh->do($sql, undef, $new_codes, time, $pause->{User}{userid}, $u->{userid})
+                              or push @{$pause->{ERROR}}, sprintf(qq{Could not enter the data into the database: <i>%s</i>.},$dbh->errstr);
```

Comment:
Probably it's fine, but: probably we need to escape `$dbh->errstr` for entities?

Reply:
As the PAUSE is now a Mojo app, error messages stored in the `ERROR` list should eventually be escaped at https://github.com/andk/pause/blob/master/lib/pause_2017/templates/layouts/layout.html.ep#L58 , but it's arguable if we should show this kind of raw error message to a user.

Reply:
This sort of internal error is really none of the user’s business. Normal users won’t be able to act on it anyway, so it’s no use to them, but it might be of use to an attacker, whom it might help with crafting an injection if such a vulnerability is ever present somewhere. Showing the raw error to the user is fine if the application is running in some kind of dev/debug mode, but in production it should only be logged somewhere (ideally with some kind of a request ID (which maybe already exists – apologies for not knowing)) and shown to the user only as a generic 500 “sorry, something went wrong” error screen. I don’t know how the rest of the code handles errors though (apologies again), so this comment might be about the codebase in general and not actionable within this PR.